src/audio_midiout.cpp

```diff
@@ -0,0 +1,387 @@
+	delta_step = (fade_end - volume) / fade_steps;
+}
+
+void GenericMidiOut::SetVolume(int new_volume) {
+	// cancel any pending fades
+	fade_steps = 0;
+
+	volume = new_volume / 100.0f;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+		device->SendMidiMessage(msg);
+	}
+}
+
+int GenericMidiOut::GetVolume()const {
+	if (fade_steps > 0) {
+		return static_cast<int>(fade_end * 100);
+	}
+	return static_cast<int>(volume * 100);
+}
+
+void GenericMidiOut::Pause() {
+	paused = true;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, 0);
+		device->SendMidiMessage(msg);
+	}
+}
+
+void GenericMidiOut::Resume() {
+	paused = false;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+		device->SendMidiMessage(msg);
+	}
+}
+
+void GenericMidiOut::Rewind() {
+	seq->rewind();
+}
+
+int GenericMidiOut::GetLoopCount() const {
+	return loop_count;
+}
+
+bool GenericMidiOut::IsFinished() const {
+	return seq->is_at_end();
+}
+
+bool GenericMidiOut::SetPitch(int pitch) {
+	tempo_multiplier = pitch / 100.0f;
+	return true;
+}
+
+int GenericMidiOut::GetPitch() const {
+	return static_cast<int>(tempo_multiplier * 100);
+}
+
+int GenericMidiOut::GetTicks() const {
+	assert(!tempo.empty());
+
+	return tempo.back().GetTicks(mtime);
+}
+
+void GenericMidiOut::Update(long long delta) {
+	if (paused) {
+		return;
+	}
+	if (fade_steps >= 0 && mtime - last_fade_mtime > 0.1f) {
+		volume = std::max(0.0f, std::min(1.0f, volume + delta_step));
+		for (int i = 0; i < 16; i++) {
+			uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+			device->SendMidiMessage(msg);
+		}
+		last_fade_mtime = mtime;
+		fade_steps -= 1;
+	}
+
+	seq->play(mtime, this);
+	mtime = mtime + ((delta / 1000000.0f) * tempo_multiplier);
+
+	if (IsFinished() && looping) {
+		mtime = seq->rewind_to_loop();
+		reset_tempos_after_loop();
+	}
+}
+
+bool GenericMidiOut::Seek(std::streamoff offset, std::ios_base::seekdir origin) {
+	if (offset == 0 && origin == std::ios_base::beg) {
+		seq->rewind();
+		mtime = 0.0f;
+		reset_tempos_after_loop();
+		SendMessageToAllChannels(midimsg_all_note_off(0));
+		return true;
+	}
+	return false;
+}
+
+void GenericMidiOut::Reset() {
+	// Generate a MIDI reset event so the device doesn't
+	// leave notes playing or keeps any state
+	reset();
+}
+
+void GenericMidiOut::SendMessageToAllChannels(uint32_t midi_msg) {
+	for (int channel = 0; channel < 16; channel++) {
+		uint8_t event_type = (midi_msg & 0x0000F0) >> 4;
+		midi_msg |= (((event_type << 4) & 0xF0) | (channel & 0x0F)) & 0x0000FF;
+		device->SendMidiMessage(midi_msg);
+	}
+}
+
+void GenericMidiOut::midi_message(int, uint_least32_t message) {
+	uint8_t event_type = (message & 0x0000F0) >> 4;
+	uint8_t channel = (message & 0x00000F);
+	uint8_t value1 = (message & 0x00FF00) >> 8;
+	uint8_t value2 = (message & 0xFF0000) >> 16;
```

Comment:
It can be worth adding static inline helper functions to grab the event_type, channel, and values from the uint32_t. Then you don't need to copy paste the code.